rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+  d = d.map(parser_fn)
+
+  assert effective_batch_size % ctx.num_replicas_in_sync == 0
+  return d.batch(effective_batch_size // ctx.num_replicas_in_sync)
+
+```
+
+### GAN
+
+Below is a usage example for a GAN which uses two optimizers in the step
+function.
+
+```python
+def sample_noise(batch_size):
+  return tf.truncated_normal(
+      shape=(batch_size, num_latents), mean=0.0, stddev=1.0)
+
+def input_fn(ctx):
+  assert effective_batch_size % ctx.num_replicas_in_sync == 0
+  batch_size = effective_batch_size // ctx.num_replicas_in_sync
+  ds = cifar.Cifar10(batch_size)
+  return ds.map(lambda x: (x['image'], sample_noise(batch_size)))
+
+with replicator.scope():
+  discriminator = GoodfellowDiscriminator(DefaultDiscriminator2D())
+  generator = DefaultGenerator2D()
+  gan = GAN(discriminator, generator)
+  disc_optimizer = tf.train.AdamOptimizer(disc_learning_rate, beta1=0.5, beta2=0.9)
+  gen_optimizer = tf.train.AdamOptimizer(gen_learning_rate, beta1=0.5, beta2=0.9)
+
+def discriminator_step(ctx, inputs):
+  del ctx  # Unused.
+  image, noise = inputs
+  gan_output = gan.connect(image, noise)
+  disc_loss, disc_vars = gan_output.discriminator_loss_and_vars()
+  disc_train_op = disc_optimizer.minimize(disc_loss, var_list=disc_vars)
+
+  with tf.control_dependencies([disc_train_op]):
+    return tf.identity(disc_loss)
+
+def generator_step(ctx, inputs):
+  del ctx  # Unused.
+  image, noise = inputs
+  gan_output = gan.connect(image, noise)
+  gen_loss, gen_vars = gan_output.generator_loss_and_vars()
+  gen_train_op = gen_optimizer.minimize(gen_loss, var_list=gen_vars)
+
+  with tf.control_dependencies([gen_train_op]):
+    return tf.identity(gen_loss)
+
+inputs = replicator.prepare_input(input_fn)
+per_replica_disc_losses = replicator.run(discriminator_step, inputs)
+per_replica_gen_losses = replicator.run(generator_step, inputs)
+mean_disc_loss = tf.reduce_mean(tf.stack(per_replica_disc_losses))
+mean_gen_loss = tf.reduce_mean(tf.stack(per_replica_gen_losses))
+
+with tf.Session() as session:
+  session.run(replicator.initialize())
+  for _ in xrange(num_train_steps):
+    for _ in xrange(num_disc_steps):
+      disc_loss = session.run(mean_disc_loss)
+    for _ in xrange(num_gen_steps):
+      gen_loss = session.run(mean_gen_loss)
+  session.run(replicator.finalize())
+```
+
+### Reinforcement Learning
+
+This is an example of
+[IMPALA](https://deepmind.com/blog/impala-scalable-distributed-deeprl-dmlab-30/)-like
+Reinforcement Learning system, converted to eager mode.
+
+```python
+tf.enable_eager_execution()
+
+with replicator.scope():
+  agent = Agent(num_actions, hidden_size, entropy_cost, baseline_cost)
+  optimizer = tf.train.RMSPropOptimizer(learning_rate)
+
+# Queues of trajectories from actors.
+queues = []
+def learner_input(ctx):
+  del ctx  # Unused.
+  queue = tf.FIFOQueue(
+      capacity=1, dtypes=trajectory_dtypes, shapes=trajectory_shapes)
+  queues.append(queue)
+
```

Comment:
It feels strange to be showing deprecated-in-1.x code as a part of this review for a 2.0 interface.

Reply:
That is correct, the queues are not deprecated, just QueueRunners.